src/thunk.rs

```diff
@@ -0,0 +1,286 @@
+
+enum ThunkVal<'a, T> {
+    Deferred(ThunkFn<'a, T>),
+    Materialized(T),
+}
+
+impl<'a, T> Debug for ThunkVal<'a, T>
+where
+    T: Debug,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        use ThunkVal::*;
+        match self {
+            Deferred(_) => write!(f, "ThunkVal::Deferred(...)"),
+            Materialized(value) => write!(f, "ThunkVal::Materialized({:?}", value),
+        }
+    }
+}
+
+/// A simple, potentially deferred or owned value.
+///
+/// The thunk can be in one of three states:
+/// * [**Deferred**](ThunkState::Deferred): The value has not yet been evaluated.
+/// * [**Materialize**](ThunkState::Materialized): The value has been evaluated and is owned.
+/// * [**Error**](ThunkState::Error): The value can never be materialized.
+///   This can happen when [`Thunk::memoize`] is used to *attempt* to materialize a deferred
+///   value in place.
+#[derive(Debug)]
+pub struct Thunk<'a, T>(IonResult<ThunkVal<'a, T>>);
+
+const EMPTY_THUNK_ERROR_TEXT: &'static str = "Empty thunk";
+
+impl<'a, T> Thunk<'a, T> {
+    #[inline]
+    pub fn wrap(value: T) -> Thunk<'static, T> {
+        Thunk(Ok(ThunkVal::Materialized(value)))
+    }
+
+    #[inline]
+    pub fn defer<F>(closure: F) -> Thunk<'a, T>
+    where
+        F: FnMut() -> IonResult<T> + 'a,
+    {
+        Thunk(Ok(ThunkVal::Deferred(Box::new(closure))))
+    }
+
+    /// Evaluates the thunk, consuming it and returning the underlying value.
+    pub fn evaluate(self) -> IonResult<T> {
+        use ThunkVal::*;
+        match self.0 {
+            Ok(Deferred(mut func)) => func(),
+            Ok(Materialized(value)) => Ok(value),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Evaluates the deferred value and returns it as a thunk.
+    pub fn materialize(self) -> IonResult<Thunk<'static, T>> {
+        use ThunkVal::*;
+        match self.0 {
+            Ok(Deferred(mut func)) => {
+                let value = func()?;
+                Ok(Thunk::wrap(value))
+            }
+            Ok(Materialized(value)) => Ok(Thunk::wrap(value)),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Evaluates in place and replaces deferred value with a materialized one.
+    ///
+    /// If the underlying deferred value is an error, the state of the thunk is in error.
+    /// An error state can be thought of as a deferred value that will never happen.
+    pub fn memoize(&mut self) -> IonResult<&T> {
+        use ThunkVal::*;
+        let value_res = self.remove();
+        // move in the new, materialized value
+        match value_res {
+            Ok(val) => self.0 = Ok(Materialized(val)),
+            Err(e) => self.0 = Err(e),
+        };
+        // at this point we can now return a reference to our internal materialized value
+        // or the error
+        match &self.0 {
+            Ok(Deferred(_)) => {
+                // XXX should not be possible
+                panic!("Thunk memoization is still deferred")
```

Comment:
I like [`unreachable!`](https://doc.rust-lang.org/std/macro.unreachable.html) for this use case.

Reply:
Updated as per suggestion.